project-base/storefront/components/Pages/BrandDetail/helpers.ts

```diff
@@ -0,0 +1,113 @@
+import { DEFAULT_PAGE_SIZE } from 'components/Blocks/Pagination/Pagination';
+import { getEndCursor } from 'components/Blocks/Product/Filter/helpers/getEndCursor';
+import {
+    ProductOrderingModeEnumApi,
+    Maybe,
+    ProductFilterApi,
+    BrandProductsQueryApi,
+    BrandProductsQueryVariablesApi,
+    BrandProductsQueryDocumentApi,
+} from 'graphql/generated';
+import { mapParametersFilter } from 'helpers/filterOptions/mapParametersFilter';
+import {
+    getPreviouslyQueriedProductsFromCache,
+    mergeNewProductsWithPreviouslyQueried,
+} from 'helpers/pagination/loadMore';
+import { getUrlWithoutGetParameters } from 'helpers/parsing/getUrlWithoutGetParameters';
+import { useQueryParams } from 'hooks/useQueryParams';
+import { useRouter } from 'next/router';
+import { useRef, useState, useEffect } from 'react';
+import { useClient, Client } from 'urql';
+import { getSlugFromUrl } from 'utils/getSlugFromUrl';
+
+export const useBrandProductsData = (): [undefined | BrandProductsQueryApi, boolean, boolean] => {
+    const client = useClient();
+    const { asPath } = useRouter();
+    const { filter, sort, currentPage, currentLoadMore } = useQueryParams();
+    const previousLoadMoreRef = useRef(currentLoadMore);
+
+    const urlSlug = getSlugFromUrl(getUrlWithoutGetParameters(asPath));
+    const mappedFilter = mapParametersFilter(filter);
+
+    const [brandProductsData, setBrandProductsData] = useState<undefined | BrandProductsQueryApi>(
+        readBrandProductsFromCache(
+            client,
+            urlSlug,
+            sort,
+            mappedFilter,
+            getEndCursor(currentPage),
+            DEFAULT_PAGE_SIZE * (currentLoadMore + 1),
+        ),
+    );
+    const [fetching, setFetching] = useState(false);
+    const [loadMoreFetching, setLoadMoreFetching] = useState(false);
+
+    useEffect(() => {
+        if (previousLoadMoreRef.current === currentLoadMore || currentLoadMore === 0) {
+            setFetching(true);
+        } else {
+            setLoadMoreFetching(true);
+            previousLoadMoreRef.current = currentLoadMore;
+        }
+
+        const previouslyQueriedProductsFromCache = getPreviouslyQueriedProductsFromCache(
+            client,
+            urlSlug,
+            sort,
+            mappedFilter,
+            DEFAULT_PAGE_SIZE,
+            currentPage,
+            currentLoadMore,
+            readBrandProductsFromCache,
+        );
+
+        const pageSize = previouslyQueriedProductsFromCache
+            ? DEFAULT_PAGE_SIZE
+            : DEFAULT_PAGE_SIZE * (currentLoadMore + 1);
+        const isMoreThanOnePage = pageSize > DEFAULT_PAGE_SIZE;
+        const endCursor = isMoreThanOnePage ? getEndCursor(currentPage) : getEndCursor(currentPage, currentLoadMore);
```

Comment:
I am not sure if this is better readable, but I will leave on you

```suggestion
        const endCursor = getEndCursor(currentPage, isMoreThanOnePage ? undefined :  currentLoadMore);
```

The same would be applied to other `helpers`.

Reply:
No problem :)